### PR TITLE
Proposal discussion state bloat fix

### PR DESCRIPTION
**Pull request by conectado**
# Fix Proposal Discussion for #2196

[Applies to proposal discussion this comments.](https://github.com/Joystream/joystream/issues/2196#issuecomment-802111952)

Also fixes Joystream/audits#13

### CODE REVIEW

File: runtime-modules/proposals/discussion/Cargo.toml [hunk 1]
```diff
@@ -11,21 +11,19 @@ sp-std = { package = 'sp-std', default-features = false, git = 'https://github.c
 frame-support = { package = 'frame-support', default-features = false, git = 'https://github.com/paritytech/substrate.git', rev = 'a200cdb93c6af5763b9c7bf313fa708764ac88ca'}
 frame-system = { package = 'frame-system', default-features = false, git = 'https://github.com/paritytech/substrate.git', rev = 'a200cdb93c6af5763b9c7bf313fa708764ac88ca'}
 common = { package = 'pallet-common', default-features = false, path = '../../common'}
+balances = { package = 'pallet-balances', default-features = false, git = 'https://github.com/paritytech/substrate.git', rev = 'a200cdb93c6af5763b9c7bf313fa708764ac88ca'}
```
Comment (shamil-gadelshin) [inline runtime-modules/proposals/discussion/Cargo.toml:14]: Please, add these crates to the 'std' feature list.

File: runtime/src/lib.rs [hunk 2]
```diff
@@ -836,6 +836,10 @@ impl Default for Call {
 
 parameter_types! {
     pub const MaxWhiteListSize: u32 = 20;
+    pub const ProposalsPostDeposit: Balance = 2000;
+    // module : proposals_discussion
+    pub const ProposalsDiscussionModuleId: ModuleId = ModuleId(*b"mo:propo");
```
Comment (shamil-gadelshin) [inline runtime/src/lib.rs]: Please, change the module id to reflect the name of the proposals discussion module, not just proposals.

File: runtime-modules/proposals/discussion/src/lib.rs [hunk 3]
```diff
@@ -214,21 +240,85 @@ decl_module! {
...
+                <PostThreadIdByPostId<T>>::insert(thread_id, post_id, new_post);
+            }
 
-            let post_id = T::PostId::from(new_post_id);
-            <PostThreadIdByPostId<T>>::insert(thread_id, post_id, new_post);
             PostCount::put(next_post_count_value);
             Self::deposit_event(RawEvent::PostCreated(post_id, post_author_id, thread_id, text));
        }
 
+        #[weight = WeightInfoDiscussion::<T>::delete_post()]
```
Comment (shamil-gadelshin) [inline runtime-modules/proposals/discussion/src/lib.rs:291]: Please, change the module comment to reflect the new extrinsic (and other changes if any).

File: runtime-modules/proposals/discussion/src/tests/mock.rs [hunk 4]
```diff
@@ -9,7 +9,7 @@ use sp_core::H256;
 use sp_runtime::{
```
Comment (shamil-gadelshin) [inline runtime-modules/proposals/discussion/src/tests/mock.rs:9]: Shouldn't we change the frame_system::AccountId type to u128 after introducing `module_id` based accounts?

File: runtime-modules/proposals/discussion/src/lib.rs [hunk 5]
```diff
@@ -260,6 +350,9 @@ decl_module! {
 
             // mutation
 
+            let mut new_post = <PostThreadIdByPostId<T>>::get(thread_id, post_id);
```
Comment (shamil-gadelshin) [inline runtime-modules/proposals/discussion/src/lib.rs]: Please, change to the "mutate" pattern.

File: runtime-modules/proposals/discussion/src/benchmarking.rs [hunk 6]
```diff
@@ -306,6 +308,40 @@ benchmarks! {
...
     }
 
+    delete_post {
+        // We do this to ignore the id 0 because the `Test` runtime
+        // returns 0 as an invalid id but 1 as a valid one
+        let (_, _) = member_account::<T>("caller_member", 0);
+        let (account_id, caller_member_id) = member_account::<T>("caller_member", 1);
+
+        // Worst case scenario there is a council
+        elect_council::<T>(2);
```
Comment (shamil-gadelshin) [inline runtime-modules/proposals/discussion/src/benchmarking.rs]: I failed to grasp the council effect on the worst-case here. Do you mind commenting on it, please?

**Review by shamil-gadelshin — CHANGES_REQUESTED** [review]
Looks OK. Left several mostly cosmetic issues.

File: runtime-modules/proposals/discussion/src/benchmarking.rs (hunk 6 above)
Comment (conectado) [reply runtime-modules/proposals/discussion/src/benchmarking.rs]: This was a left over it was not needed.

Comment (shamil-gadelshin) [reply runtime-modules/proposals/discussion/src/benchmarking.rs]: Ok. There are other such code lines in this benchmarking file. Remove them too, please, if necessary.

File: runtime-modules/proposals/discussion/Cargo.toml [hunk 7]
```diff
@@ -47,4 +44,5 @@ std = [
 	'frame-support/std',
 	'frame-system/std',
     'common/std',
+    'balances/std',
```
Comment (shamil-gadelshin) [inline runtime-modules/proposals/discussion/Cargo.toml:47]: please, add also `sp-runtime` crate

**Review by shamil-gadelshin — CHANGES_REQUESTED** [review]
Two issues still present:
- `sp_runtime` entry in the Cargo.toml is missing in the `std` feature list
- `delete_post` is missing in the extrinsic list of the module comment

**Review by shamil-gadelshin — APPROVED** [review]
LGTM

**Pull request merged**